### PR TITLE
Better example for W0640: cell-var-from-loop

### PR DESCRIPTION
**Pull request by SubaruArai**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :scroll: Docs          |

## Description

Add another example for W0640: cell-var-from-loop.
Example to use `functools.partial`

### CODE REVIEW

**Comment by SubaruArai** [comment]
I felt this is trivial, so left the news fragment untouched.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for the contribution :) Could you create a good/ directory with two aptly named files inside instead please ? The final result will be more readable. You can look at bad-chained-comparison for an example of what to do.

**Review by Pierre-Sassoulas — APPROVED** [review]
Great first contribution :) ! I'm going to remove the foo/bar in the example directly in this MR, sorry for the noise.

**Comment by Pierre-Sassoulas** [comment]
Locally I get a cell-var-from-loop from the good example, do you also have that ? (``cd doc/; pytest -k cell-var``)

**Comment by SubaruArai** [comment]
My bad, a quick look does seem that my code is wrong XD

**Comment by SubaruArai** [comment]
Fixed the example, but it seems the "bad" directory needs 2 examples.

@Pierre-Sassoulas Any idea to circumvent? Maybe I need to come up with another bad case?

**Comment by Pierre-Sassoulas** [comment]
I'm going to fix the template to be able to have one bad and two good, it's still very new :)

**Comment by Pierre-Sassoulas** [comment]
Hello, the problem do not come from the helper for the templates with directory, the good.py example actually raises cell-var-from-loop, do you mind having a look ? I'm going to fix the messages but it's not super easy to have an accurate message considering we do not have the information of the file raising a message with the current structure of the code.

**Comment by SubaruArai** [comment]
This might be a bug in pylint itself, since the code runs just fine...?

**Comment by SubaruArai** [comment]
Anyways, I've fixed the documentation to not raise an error.

**Comment by SubaruArai** [comment]
...Although technically this implementation doesn't do lazy evaluation.

**Comment by SubaruArai** [comment]
I'm starting to question if this example (print a modified string, but make it a function) is good, in the meaning that this code seems unrealistic or a bad practice. Especially if we avoid `foo/bar` to be as newcomers friendly as possible.

Also, the "bad" example runs just fine - since it redefines the function every time and calls it within the loop - before `name` changes.

The [related](https://stackoverflow.com/q/25314547/17094594) link is about `sort`, which requires a callable.
But then again, "a function that requires an callable as argument" itself is sort of an advanced concept - what would be a good example for this?

**Comment by SubaruArai** [comment]
What if we changed the name to something like `lazy_greeting`?

```python
def teacher_greeting(names):
    for name in names:

        greet = functools.partial(print, f"Hello, {name}!")

        greet()


teacher_greeting(["Graham", "John", "Terry", "Eric", "Terry", "Michael"])

**Comment by SubaruArai** [comment]
Sorry, closed by accident

**Comment by SubaruArai** [comment]
If the intend of the example was:
```python
def greet_teachers(names):
    for name in names:
        def greet():
            # do something
            print(f"Hello, {name}!")
        greet()
```

Shouldn't the good example be like:
```python
def greet_teachers(names):
    def greet(name):
        # do something
        print(f"Hello, {name}!")

    for name in names:
        greet(name)
```

**Comment by DanielNoord** [comment]
> If the intend of the example was:
> 
> ```python
> def greet_teachers(names):
>     for name in names:
>         def greet():
>             # do something
>             print(f"Hello, {name}!")
>         greet()
> ```
> 
> Shouldn't the good example be like:
> 
> ```python
> def greet_teachers(names):
>     def greet(name):
>         # do something
>         print(f"Hello, {name}!")
> 
>     for name in names:
>         greet(name)
> ```
> 
> Why make the `greet` function global?
> 
> Also, `functools.partial` will be inappropriate for this purpose.

I think these examples are very good indeed! You could consider using those?

**Comment by SubaruArai** [comment]
@DanielNoord While I definitely can, I don't like the fact that the "bad" example works just as fine - albeit less efficient.

What would be a good "bad" example?

**Comment by DanielNoord** [comment]
I think the bad example is also fine? It's the same as what we currently have in the documentation and it is the code that raises the warning. It's just bad practice to write Python code like that.

**Comment by Pierre-Sassoulas** [comment]
I noticed that the bad example produce a warning but does not actually produce a problem in what is printed (python bad.py produce the expected list of names). Ideally we would use something that is not a false positive in the doc :smile:

**Comment by SubaruArai** [comment]
It will be quite a lot of change, but what about this:

Bad:
```python
def greet_teachers(names):
    greetings = []
    for name in names:

        def greet():
            # do stuff
            print(f"hello, {name}")

        if name.isalpha():
            greetings.append(greet)

    for greet in greetings:
        greet()


greet_teachers(["Alice", "Bob", "Charlie", "Not-A-Name"])
# "hello, Not-A-Name" * 3
```

Good:
```python
def greet_teachers(names):
    names_to_greet = []

    def greet(name):
        # do something
        print(f"hello, {name}")

    for name in names:
        if name.isalpha():
            names_to_greet.append(name)

    for name in names_to_greet:
        greet(name)


greet_teachers(["Alice", "Bob", "Charlie", "Not-A-Name"])
# "hello, Alice", ...
```

**Comment by SubaruArai** [comment]
And using `functools.partial`:
```python
import functools


def greet_teachers(names):
    greetings = []
    for name in names:
        if name.isalpha():
            greetings.append(functools.partial(print, f"hello, {name}"))

    for greet in greetings:
        greet()


greet_teachers(["Alice", "Bob", "Charlie", "Not-A-Name"])
# "hello, Alice", ...
```

**Comment by Pierre-Sassoulas** [comment]
Sound good ! I was hoping something simpler was possible, but I did not have the time to find a simple example and a working example is paramount.

**Comment by SubaruArai** [comment]
While this does lazy formatting, pylint will complain to use f-strings:
```python
import functools


def greet_teachers(names):
    greeting_msgs = []
    for name in names:
        if name.isalpha():
            greeting_msgs.append(functools.partial("hello, {}".format, name))

    for greeting_msg in greeting_msgs:
        print(greeting_msg())


greet_teachers(["Alice", "Bob", "Charlie", "Not-A-Name"])
# "hello, Alice", ...
```

**Comment by SubaruArai** [comment]
> I was hoping something simpler was possible

Yeah, I feel that too. The thing is, having this warning mostly means you're abusing lazy evaluation - which is kind of complicated for newcomers anyways.

**Comment by SubaruArai** [comment]
Ok, I've updated the pr.

File: doc/data/messages/c/cell-var-from-loop/bad.py [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
+
+    for greet in greetings:
+        # the "name" variable is evaluated when the function is called here,
+        # which is the last value it had in the loop - "Not-A-Name"
         greet()
 
 
-teacher_greeting(["Graham", "John", "Terry", "Eric", "Terry", "Michael"])
+teacher_greeting(["Alice", "Bob", "Charlie", "Not-A-Name"])
+# "Hello, Not-A-Name!" * 3
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/c/cell-var-from-loop/bad.py:19]:
```suggestion
teacher_greeting(["Graham", "John", "Terry", "Eric", "Terry", "Michael"])
# "Hello, Michael"
# "Hello, Michael"
# "Hello, Michael"
# "Hello, Michael"
# "Hello, Michael"
```

File: doc/data/messages/c/cell-var-from-loop/good/functools.partial.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            greetings.append(functools.partial(print, f"Hello, {name}!"))
+
+    for greet in greetings:
+        # `partial`s are called like functions, but you've already passed the
+        # arguments to them
+        greet()
+
+
+greet_teachers(["Alice", "Bob", "Charlie", "Not-A-Name"])
+# "Hello, Alice!", ...
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/c/cell-var-from-loop/good/functools.partial.py:19]:
```suggestion
teacher_greeting(["Graham", "John", "Terry", "Eric", "Terry", "Michael"])
# "Hello, Graham"
# "Hello, John"
# "Hello, Eric"
# "Hello, Terry"
# "Hello, Michael"
```
This is a Monthy Python reference, let's keep it :p

File: doc/data/messages/c/cell-var-from-loop/good/new_function.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+        print(f"Hello, {name}!")
+
+    for name in names:
+        if name.isalpha():
+            # we're passing the value of "name" to the function here
+            greet(name)
+
+
+teacher_greeting(["Alice", "Bob", "Charlie", "Not-A-Name"])
+# "Hello, Alice!", ...
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/c/cell-var-from-loop/good/new_function.py:13]:
```suggestion
teacher_greeting(["Graham", "John", "Terry", "Eric", "Terry", "Michael"])
# "Hello, Graham"
# "Hello, John"
# "Hello, Eric"
# "Hello, Terry"
# "Hello, Michael"
```
This is a Monthy Python reference, let's keep it :p

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
LGTM, it's blocked by https://github.com/pylint-dev/pylint/pull/8991 though, let's rebase on it when it's merged.

File: doc/data/messages/c/cell-var-from-loop/good/new_function.py (hunk 3 above)
Comment (SubaruArai) [reply doc/data/messages/c/cell-var-from-loop/good/new_function.py:13]: Aha, of course!

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for taking the time to make the doc about this message great ! (Also permitted to make the template better in the process 👌 )

**Comment by Pierre-Sassoulas** [comment]
Congratulation on becoming a contributor, next time you won't have to wait for approval to launch a CI job :)

**Comment by SubaruArai** [comment]
No, thank you all for the patience!

**Pull request merged**